lib/PrivateAppUserTokenManager.ts

```diff
@@ -0,0 +1,249 @@
+        messageDetail = err.message;
+      }
+      logger.warn(
+        i18n(`${i18nKey}.errors.apiError`, {
+          accountId: this.accountId,
+          appId: appId,
+          messageDetail: messageDetail,
+        })
+      );
+    }
+  }
+
+  isEnabled(): boolean {
+    return this.enabled;
+  }
+
+  private setCacheAndRefresh(
+    appId: number,
+    token: PrivateAppUserTokenResponse,
+    scopeGroups: string[]
+  ) {
+    if (token === undefined || token === null) {
+      logger.warn(
+        i18n(`${i18nKey}.errors.refreshFailed`, {
+          accountId: this.accountId,
+          appId: appId,
+        })
+      );
+      return;
+    }
+    this.tokenMap.set(appId, token);
+    if (this.tokenMapIntervalId.has(appId)) {
+      clearInterval(this.tokenMapIntervalId.get(appId));
+    }
+    const now = moment.utc();
+    const refreshDelayMillis = Math.max(
+      moment
+        .utc(token.expiresAt)
+        .subtract(5, 'minutes')
+        .diff(now, 'milliseconds'),
+      0
+    );
+    this.tokenMapIntervalId.set(
+      appId,
+      setInterval(
+        () => this.refreshToken(appId, token.userTokenKey, scopeGroups),
+        refreshDelayMillis
+      )
+    );
+    logger.debug(
+      i18n(`${i18nKey}.refreshScheduled`, {
+        appId: appId,
+        expiresAt: token.expiresAt,
+        refreshTime: now.add(refreshDelayMillis, 'milliseconds').toISOString(),
+      })
+    );
+  }
+
+  private async createOrGetActiveToken(
+    appId: number,
+    scopeGroups: string[]
+  ): Promise<PrivateAppUserTokenResponse> {
+    const maybeToken = await this.getExistingToken(appId);
+    if (maybeToken === null) {
+      logger.debug(
+        i18n(`${i18nKey}.create`, {
+          appId: appId,
+        })
+      );
+      return createPrivateAppUserToken(this.accountId, appId, scopeGroups);
+    } else if (
+      moment
+        .utc()
+        .add(5, 'minutes')
+        .isAfter(moment.utc(maybeToken.expiresAt)) ||
+      !PrivateAppUserTokenManager.doesTokenHaveAllScopes(
+        maybeToken,
+        scopeGroups
+      )
+    ) {
+      logger.debug(
+        i18n(`${i18nKey}.refresh`, {
+          appId: appId,
+        })
+      );
+      return updatePrivateAppUserToken(
+        this.accountId,
+        appId,
+        maybeToken.userTokenKey,
+        scopeGroups
+      );
+    }
+    return maybeToken;
+  }
+
+  private static doesTokenHaveAllScopes(
+    token: PrivateAppUserTokenResponse | undefined,
+    requestedScopes: string[]
+  ): boolean {
+    if (token === undefined) {
+      return false;
+    }
+    return requestedScopes.every(scopeGroup =>
+      token.scopeGroups.includes(scopeGroup)
+    );
+  }
+
+  private async getExistingToken(
+    appId: number
+  ): Promise<PrivateAppUserTokenResponse | null> {
+    try {
+      return await fetchPrivateAppUserToken(this.accountId, appId);
+    } catch (err) {
+      if (err instanceof AxiosError) {
+        const axiosError = err as AxiosError;
```

Comment:
If we use `if(isAxiosError(err))` it **_should_** type narrow and we can ditch the type assertion `err as AxiosError`